R/mock-session.R

```diff
@@ -0,0 +1,281 @@
+
+        # timerCallbacks must run before flushReact.
+        private$timer$executeElapsed()
+        private$flush()
+      }
+
+      private$timer$elapse(msLeft)
+
+      # Run again in case our callbacks resulted in a scheduled
+      # function that needs executing.
+      private$timer$executeElapsed()
+      private$flush()
+    },
+
+    .now = function() {
+      # Contract is to return Sys.time, which is seconds, not millis.
+      private$timer$getElapsed()/1000
+    },
+
+    defineOutput = function(name, value, label) {
+      obs <- observe({
+        # We could just stash the promise, but we get an "unhandled promise error". This bypasses
+        prom <- NULL
+        tryCatch({
+          v <- value(self, name)
+          if (!promises::is.promise(v)){
+            # Make our sync value into a promise
+            prom <- promises::promise(function(resolve, reject){ resolve(v) })
+          } else {
+            prom <- v
+          }
+        }, error=function(e){
+          # Error running value()
+          prom <<- promises::promise(function(resolve, reject){ reject(e) })
+        })
+
+        private$outs[[name]]$promise <- hybrid_chain(
+          prom,
+          function(v){
+            list(val = v, err = NULL)
+          }, catch=function(e){
+            list(val = NULL, err = e)
+          })
+      })
+      private$outs[[name]] <- list(obs = obs, func = value, promise = NULL)
+    },
+
+    getOutput = function(name) {
+      # Unlike the real outputs, we're going to return the last value rather than the unevaluated function
+      if (is.null(private$outs[[name]]$promise)) {
+        stop("The test referenced an output that hasn't been defined yet: output$", name)
+      }
+      # Make promise return
+      v <- extract(private$outs[[name]]$promise)
+      if (!is.null(v$err)){
+        stop(v$err)
+      } else {
+        v$val
+      }
+    },
+
+    registerDataObj = function(name, data, filterFunc) {},
+    allowReconnect = function(value) {},
+    reload = function() {},
+    resetBrush = function(brushId) {
+      warning("session$brush isn't meaningfully mocked on the MockShinySession")
+    },
+    sendCustomMessage = function(type, message) {},
+    sendBinaryMessage = function(type, message) {},
+    sendInputMessage = function(inputId, message) {},
+    setBookmarkExclude = function(names) {
+      warning("Bookmarking isn't meaningfully mocked in MockShinySession")
+    },
+    getBookmarkExclude = function() {
+      warning("Bookmarking isn't meaningfully mocked in MockShinySession")
+    },
+    onBookmark = function(fun) {
+      warning("Bookmarking isn't meaningfully mocked in MockShinySession")
+    },
+    onBookmarked = function(fun) {
+      warning("Bookmarking isn't meaningfully mocked in MockShinySession")
+    },
+    doBookmark = function() {
+      warning("Bookmarking isn't meaningfully mocked in MockShinySession")
+    },
+    onRestore = function(fun) {},
```

Comment:
I think `onRestore` and `onRestored` should give warnings as well.

Reply:
This may just fold into the conversation on the other comment, but my logic for not warning here is that -- in my head -- people wouldn't expect onRestore/d to be invoked ever since there's no client to do the restoring. So defining an `onRestore` method in a module isn't harmful or broken -- it's just rightly a no-op.